src/utils/source-search.js

```diff
@@ -28,10 +28,20 @@ function getSearchState(cm) {
  * @memberof utils/source-search
  * @static
  */
-function getSearchCursor(cm, query, pos) {
-  // If the query string is all lowercase, do a case insensitive search.
-  return cm.getSearchCursor(query, pos,
-    typeof query == "string" && query == query.toLowerCase());
+function getSearchCursor(
+  cm, query, pos, caseSensitive, wholeWord, regexMatch) {
+  let moddedQuery = query;
+  if (regexMatch && query !== "") {
+    moddedQuery = new RegExp(query, caseSensitive ? "" : "i");
+  } else if (query !== "" && (wholeWord || (wholeWord && regexMatch))) {
+    moddedQuery = new RegExp(`\\b${query}\\b`, caseSensitive ? "" : "i");
+  }
+
+  if (query !== "" && typeof moddedQuery == "string") {
+    return cm.getSearchCursor(moddedQuery, pos, !caseSensitive);
+  }
+
+  return cm.getSearchCursor(moddedQuery, pos);
```

Comment:
this looks for a candidate where you might want to use the spy pattern https://github.com/mochajs/mocha/wiki/Spies and pass in a spy or stub function for `cm`.

I could see some unit tests really helping here :) 

as you said earlier